src/snp.rs

```diff
@@ -0,0 +1,262 @@
+            "genoa" => Ok(Self::Genoa),
+            _ => Err(AttesterError::InvalidTee),
+        }
+    }
+}
+
+impl SnpGeneration {
+    fn to_lower(&self) -> String {
+        match self {
+            SnpGeneration::Milan => "milan".to_string(),
+            SnpGeneration::Genoa => "genoa".to_string(),
+        }
+    }
+
+    fn to_proper(&self) -> String {
+        match self {
+            SnpGeneration::Milan => "Milan".to_string(),
+            SnpGeneration::Genoa => "Genoa".to_string(),
+        }
+    }
+}
+
+pub struct CaChain {
+    pub ark: X509,
+    pub ask: X509,
+}
+
+impl CaChain {
+    pub fn get(gen: &SnpGeneration) -> Self {
+        let mut ark_buf: Vec<u8> = Vec::new();
+        let mut ask_buf: Vec<u8> = Vec::new();
+
+        let gen_str = gen.to_lower();
+
+        let mut ark_file = File::open(&format!("certs/{}/ark.pem", gen_str)).unwrap();
+        let mut ask_file = File::open(&format!("certs/{}/ask.pem", gen_str)).unwrap();
+
+        ark_file.read_to_end(&mut ark_buf).unwrap();
+        ask_file.read_to_end(&mut ask_buf).unwrap();
+
+        let ark = X509::from_pem(&ark_buf[..]).unwrap();
+        let ask = X509::from_pem(&ask_buf[..]).unwrap();
+
+        Self { ark, ask }
+    }
+
+    pub fn pkeys(&self) -> (PKey<Public>, PKey<Public>) {
+        (
+            self.ark.public_key().unwrap(),
+            self.ask.public_key().unwrap(),
+        )
+    }
+}
+
+pub struct SnpAttester {
+    workload_id: String,
+    nonce: String,
+    tee_config: String,
+    rsa: Option<Rsa<Public>>,
+}
+
+impl SnpAttester {
+    pub fn new(workload_id: String, nonce: String, tee_config: String) -> Self {
+        Self {
+            workload_id,
+            nonce,
+            tee_config,
+            rsa: None,
+        }
+    }
+}
+
+impl Attester for SnpAttester {
+    fn workload_id(&self) -> &String {
+        &self.workload_id
+    }
+
+    fn challenge(&mut self) -> Result<Challenge, AttesterError> {
+        /*
+         * For SNP, we're really only concerned about the nonce in the
+         * Challenge. Therefore, we have no reason to create an SnpChallenge
+         * struct. Use the generic Challenge to send back to the client.
+         */
+        Ok(Challenge {
+            nonce: self.nonce.clone(),
+            extra_params: "".to_string(),
+        })
+    }
+
+    fn attest(
+        &mut self,
+        attestation: &Attestation,
+        measurement: &str,
+    ) -> Result<(), AttesterError> {
+        let snp: SnpAttestation = json::from_str(&attestation.tee_evidence).unwrap();
+        let report: AttestationReport = json::from_str(&snp.report).unwrap();
+        let pkey: &TeePubKey = &attestation.tee_pubkey;
+        if !nonce_pkey_hash(&report, pkey, &self.nonce) {
+            return Err(AttesterError::SnpNoncePubkeyHashInvalid);
+        }
+
+        if measurement.as_bytes() != report.measurement {
+            return Err(AttesterError::SnpMeasurementInvalid);
+        }
```

Comment:
```
if measurement.as_bytes() != report.measurement {
            return Err(AttesterError::SnpMeasurementInvalid);
}
```

I need to test this further with `oci2cw`.

Reply:
`oci2cw` will register the measurement encoded as en hexadecimal string, so reference-kbs needs to decode it first before the comparison:

```
+        let decoded_lm = hex::decode(measurement).unwrap();
+        println!("measurement: data={:?} report={:?}", decoded_lm, report.measurement);
+        if decoded_lm != report.measurement {
+            return Err(AttesterError::SnpMeasurementInvalid);
+        }
```

I've tested this a bit, and still didn't get a proper measurement match. We might need to update the vmsa blob too.